File: back/kotlin/app/core/domain/GameUseCases.kt
package core.domain

import org.slf4j.LoggerFactory
interface GameUseCasesB {
    suspend fun initializeGame(): Result<GameType>
    suspend fun resetGame(gameID: String): Result<GameType>
    suspend fun makeMove(gameID: String, position: CellPosition, player: PlayerType): Result<GameType>
}

class GameUseCases(private val gameRepo: GameRepository, private val gameLogic: GameLogicB): GameUseCasesB {

    companion object {
        private val logger = LoggerFactory.getLogger(GameUseCases::class.java)
    }

    override suspend fun initializeGame(): Result<GameType> {

        val newKeyResult = gameLogic.generateNewID()
        val boardResult = gameLogic.generateNewBoard()
        val playerResult = gameLogic.randomPlayer()

        when {
            newKeyResult is Result.Success<String> && boardResult is Result.Success<BoardType> && playerResult is Result.Success<PlayerType> -> {
                return gameRepo.createNewGame(newKeyResult.data, State.BoardState(boardResult.data), playerResult.data)
            }
            else -> {
                val errors = buildString {
                    if (newKeyResult !is Result.Success<String>) {
                        appendln("Error generating new ID: ${(newKeyResult as Result.Error).exception}")
                    }
                    if (boardResult !is Result.Success<BoardType>) {
                        appendln("Error generating new board: ${(boardResult as Result.Error).exception}")
                    }
                    if (playerResult !is Result.Success<PlayerType>) {
                        appendln("Error getting random player: ${(playerResult as Result.Error).exception}")
                    }
                }
                return Result.Error(IllegalStateException("Game initialization failed: $errors"))
            }
        }

    }

    override suspend fun resetGame(gameID: String): Result<GameType> {

        val boardResult = gameLogic.generateNewBoard()
        val playerResult = gameLogic.randomPlayer()


        when {
            boardResult is Result.Success<BoardType> && playerResult is Result.Success<PlayerType> -> {
                return gameRepo.resetGame(gameID, State.BoardState(boardResult.data), playerResult.data)
            }
            else -> {
                val errors = buildString {
                    if (boardResult !is Result.Success<BoardType>) {
                        appendln("Error generating new board: ${(boardResult as Result.Error).exception}")
                    }
                    if (playerResult !is Result.Success<PlayerType>) {
                        appendln("Error getting random player: ${(playerResult as Result.Error).exception}")
                    }
                }
                return Result.Error(IllegalStateException("Game initialization failed: $errors"))
            }
        }

    }

    override suspend fun makeMove(gameID: String, position: CellPosition, player: PlayerType): Result<GameType> {

        val newBoardStateResult = gameRepo.updateBoard(gameID, position, player)
        when (newBoardStateResult) {
            is Result.Success<StateType> -> {
                val data = newBoardStateResult.data
                logger.info("Newboard State Result: $data")
            }
            is Result.Error -> {
                val error = newBoardStateResult.exception
                logger.info("Newboard State Error: $error")
                return Result.Error(IllegalStateException("Error getting board: ${(newBoardStateResult as Result.Error).exception}"))
            }
            is Result.NotFound -> {
                logger.warn("Result not found")
                return Result.NotFound
            }
            else -> {
                logger.info("Case not covered")
                return Result.NotFound
            }
        }

        var gameState: GameState = GameState.IN_PROGRESS
        var winner: PlayerType? = null

        val checkWinnerAndDrawResult = gameLogic.checkForWinner(newBoardStateResult.data)
        when (checkWinnerAndDrawResult) {
            is Result.Success<GameResult> -> {
                val data = checkWinnerAndDrawResult.data
                logger.info("Winner Check Result: $data")
                if (checkWinnerAndDrawResult.data.winner != null){
                    gameState = GameState.WON
                    winner = checkWinnerAndDrawResult.data.winner
                } else if(checkWinnerAndDrawResult.data.draw){
                    gameState = GameState.DRAW
                }
            }
            is Result.Error -> {
                val error = checkWinnerAndDrawResult.exception
                logger.info("Winner Check Error: $error")
                return Result.Error(IllegalStateException("Error checking for winner: ${(checkWinnerAndDrawResult as Result.Error).exception}"))
            }
            is Result.NotFound -> {
                logger.warn("Result not found")
            }
            else -> {
                logger.info("Case not covered")
            }
        }

        val nextPlayerResult = gameLogic.getNextPlayer(player)
        when (nextPlayerResult) {
            is Result.Success<PlayerType> -> {
                val data = nextPlayerResult.data
                logger.info("Next Player Result: $data")
            }
            is Result.Error -> {
                val error = nextPlayerResult.exception
                logger.info("Next Player Error: $error")
                return Result.Error(IllegalStateException("Error checking for winner: ${(nextPlayerResult as Result.Error).exception}"))
            }
            is Result.NotFound -> {
                logger.warn("Result not found")
                return Result.NotFound
            }
            else -> {
                logger.info("Case not covered")
                return Result.NotFound
            }
        }

        return gameRepo.updateGameState(gameID, newBoardStateResult.data, GameInfo(gameState, nextPlayerResult.data, winner))
    }

}

File: back/go/app/core/domain/GameUseCases.go
package repository

import (
	"context"
	"go-ttt/app/core/data/repository"
	"go-ttt/app/core/database/redis/entity"
	"go-ttt/app/core/model"
	service "go-ttt/app/core/service/game"
)

type GameUseCases interface {
	InitializeGame(ctx context.Context) (*model.GameType, error)
	ResetGame(ctx context.Context, gameID string) (*model.GameType, error)
	MakeMove(ctx context.Context, gameID string, position *model.CellPosition, player *model.PlayerType) (*model.GameType, error)
}
type GameUseCasesImpl struct {
	gameRepo repository.GameRepository
	gameProcess service.GameLogicB
}

func NewGameUseCases(repo repository.GameRepository, logic service.GameLogicB) GameUseCases {
	return &GameUseCasesImpl{
		gameRepo: repo,
		gameProcess: logic,
	}
}

func (uc *GameUseCasesImpl) InitializeGame(ctx context.Context) (*model.GameType, error) {

	newKey, err := uc.gameProcess.GenerateNewID()
	if err != nil {
		return nil, err
	}
	board, err := uc.gameProcess.GenerateNewBoard()
	if err != nil {
		return nil, err
	}

	var boardState = model.BoardState{
		BoardType: *board,
	}

	var boardStateAsStateType model.StateType = &boardState

	player, err := uc.gameProcess.RandomPlayer()
	if err != nil {
		return nil, err
	}

	return uc.gameRepo.CreateNewGame(ctx, newKey, &boardStateAsStateType, player)
}

func (uc *GameUseCasesImpl) ResetGame(ctx context.Context, gameID string) (*model.GameType, error) {

	board, err := uc.gameProcess.GenerateNewBoard()
	if err != nil {
		return nil, err
	}

	var boardState = model.BoardState{
			BoardType: *board,
		}

	var boardStateAsStateType model.StateType = &boardState

	player, err := uc.gameProcess.RandomPlayer()
	if err != nil {
		return nil, err
	}

	return uc.gameRepo.ResetGame(ctx, gameID, &boardStateAsStateType, player)
}

func (uc *GameUseCasesImpl) MakeMove(ctx context.Context, gameID string, position *model.CellPosition, player *model.PlayerType) (*model.GameType, error) {

	newBoardState, err := uc.gameRepo.UpdateBoard(ctx, gameID, position, player)
	if err != nil {
		return nil, err
	}

	result, err := uc.gameProcess.CheckForWinner(*newBoardState)
	if err != nil {
		return nil, err
	}

	nextPlayer, err := uc.gameProcess.GetNextPlayer(player)
	if err != nil {
		return nil, err
	}

	gameState := model.InProgress
	if result.Winner != nil {
		gameState = model.Won
	}
	if(result.Draw){
		gameState = model.Draw
	}

	var gameInfo = entity.GameInfo{
		GameState: gameState,
		CurrentPlayer: *nextPlayer,
		Winner: result.Winner,
	}

	return uc.gameRepo.UpdateGameState(ctx, gameID, newBoardState, &gameInfo)
}

